atc/engine/build_step_delegate.go

```diff
@@ -373,11 +373,29 @@ func (delegate *buildStepDelegate) ConstructAcrossSubsteps(templateBytes []byte,
 		if err := yaml.Unmarshal(interpolatedBytes, &subPlan); err != nil {
 			return nil, fmt.Errorf("invalid template bytes: %w", err)
 		}
+
+		// Maps from the original subplan ID generated by the planner to the
+		// translated ID unique to the substep iteration.
+		mappedSubplanIDs := map[atc.PlanID]atc.PlanID{}
 		planIDCounter := 0
 		subPlan.Each(func(p *atc.Plan) {
-			p.ID = atc.PlanID(fmt.Sprintf("%s/%d/%d", delegate.planID, i, planIDCounter))
+			mappedID := atc.PlanID(fmt.Sprintf("%s/%d/%d", delegate.planID, i, planIDCounter))
```

Comment:
is there a specific format of the mappedID? How is it decided?

Reply:
There doesn't have to be a specific format, as long as it's unique within the build. Easiest way I thought to do that was this way - if the `across` step has plan ID `xyz`, then the `j`th subplan of the `i`th iteration gets ID `xyz/i/j`.

Note that the plan IDs don't need to be predictable, but it's kind of nice to have. The only place you'll see them is the URL when you highlight build logs in the UI, e.g. https://ci.concourse-ci.org/teams/main/pipelines/concourse/jobs/testflight/builds/879#L60f3a721/2/0:132 (here, the `across` plan ID is `60f3a721`, this is iteration with index `2`, and the subplan with index `0`).

We do a similar thing with the image check and get subplans - e.g. https://dispatcher.concourse-ci.org/teams/main/pipelines/production/jobs/bump-concourse-to-rc/builds/8#L609c9bc9/image-get:1 (the plan ID is `609c9bc9`, which we suffix with `/image-get`).